### ISSUE TITLE
print returning EINTR in 5.14

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13142]
Migrated from [rt.perl.org#119097](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=119097) (status was 'resolved')

Searchable as RT119097$

**Comment by p5pRT** [comment]
### From victor@vsespb.ru

Test case&#8203;:

\#\!/usr/bin/env perl

use strict;
use warnings;
use utf8;
use Test&#8203;::More tests => 20;
use Time&#8203;::HiRes qw/usleep ualarm/;
use IO&#8203;::Pipe;


local $SIG\{ALRM\} = sub \{ print STDERR "ALRM $$\\n" \};
my $sample = 'abxhrtf6';
my $full\_sample = 'abxhrtf6' x \(8192\-7\);
my $sample\_l = length $full\_sample;

my $ppid = $$;
my $fh = new IO&#8203;::Pipe;

if \(my $pid = fork\(\)\) \{
&nbsp;    my $child\_exited = 0;
&nbsp;    $fh\->reader\(\);
&nbsp;    $fh\->autoflush\(1\);
&nbsp;    $fh\->blocking\(1\);
&nbsp;    binmode $fh\, "&#8203;:perlio";

&nbsp;    usleep 100\_000;
&nbsp;    for \(1\.\.10\) \{
&nbsp;        my $n = read\($fh\, my $x\, $sample\_l\);
&nbsp;        die "EOF" unless $n;
&nbsp;        is $n\, $sample\_l\, "should return right amount of data";
&nbsp;        ok $x eq $full\_sample\, "should return right data"
&nbsp;    \}

&nbsp;    while\(wait\(\) \!= \-1 \)\{\};
\} else \{
&nbsp;    $fh\->writer\(\);
&nbsp;    $fh\->autoflush\(1\);
&nbsp;    $fh\->blocking\(1\);
&nbsp;    binmode $fh\, "&#8203;:perlio";

&nbsp;    for \(1\.\.10\) \{
&nbsp;        ualarm\(10\_000\);
&nbsp;        while \( \!print\($fh $full\_sample\) && $\!\{EINTR\} \) \{ print STDERR
"print failed&#8203;: \[ $\! \]\\n" \};
&nbsp;        ualarm\(0\);
&nbsp;    \}

&nbsp;    exit\(0\);
\}

1;

\_\_END\_\_


Works fine under all perls \( Linux \)\, except perl\-5\.14

when works fine&#8203;:

1\.\.20
ALRM 32265
ok 1 \- should return right amount of data
ok 2 \- should return right data
ok 3 \- should return right amount of data
ok 4 \- should return right data
ok 5 \- should return right amount of data
ok 6 \- should return right data
ok 7 \- should return right amount of data
ok 8 \- should return right data
ok 9 \- should return right amount of data
ok 10 \- should return right data
ok 11 \- should return right amount of data
ok 12 \- should return right data
ok 13 \- should return right amount of data
ok 14 \- should return right data
ok 15 \- should return right amount of data
ok 16 \- should return right data
ok 17 \- should return right amount of data
ok 18 \- should return right data
ok 19 \- should return right amount of data
ok 20 \- should return right data


in 5\.14&#8203;:

1\.\.20
ALRM 2058
print failed&#8203;: \[ Interrupted system call \]
print failed&#8203;: \[ Interrupted system call \]
ok 1 \- should return right amount of data
ok 2 \- should return right data
print failed&#8203;: \[ Interrupted system call \]
ok 3 \- should return right amount of data
ok 4 \- should return right data
print failed&#8203;: \[ Interrupted system call \]
ok 5 \- should return right amount of data
ok 6 \- should return right data
print failed&#8203;: \[ Interrupted system call \]
ok 7 \- should return right amount of data
ok 8 \- should return right data
print failed&#8203;: \[ Interrupted system call \]
ok 9 \- should return right amount of data
ok 10 \- should return right data
print failed&#8203;: \[ Interrupted system call \]
ok 11 \- should return right amount of data
ok 12 \- should return right data
print failed&#8203;: \[ Interrupted system call \]
ok 13 \- should return right amount of data
ok 14 \- should return right data
print failed&#8203;: \[ Interrupted system call \]
ok 15 \- should return right amount of data
ok 16 \- should return right data
print failed&#8203;: \[ Interrupted system call \]
ok 17 \- should return right amount of data
ok 18 \- should return right data
ok 19 \- should return right amount of data
ok 20 \- should return right data
print failed&#8203;: \[ Interrupted system call \]
^C

so print returns undef \+ EINTR\, and even if restart print operation\, it
hangs\, with both :perlio or :raw mode\.

possible similar issue&#8203;: http&#8203;://www\.perlmonks\.org/?node\_id=1026542

I wonder if this a bug\, maybe a fix should be backported\, or a test case
created\.

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Jul 31\, 2013 at 12&#8203;:30&#8203;:58PM \-0700\, Victor Efimov wrote&#8203;:
\(stuff about EINTR\)
> I wonder if this a bug\, maybe a fix should be backported\, or a test case
> created\.

This was fixed in 5\.15\.4 with the following commit\. It's unlikely to to
be back\-ported to 5\.14\.x\, since that branch is now end\-of\-life\.


commit be48bbe8d671b6841c3ec7cb734b98071afe3cd9
Author&#8203;: Chip \<chip@&#8203;pobox\.com>
Date&#8203;:   Mon Sep 19 23&#8203;:51&#8203;:49 2011 \-0700

&nbsp;    add a couple missing LEAVEs in perlio\_async\_run\(\)

\-\- 
Justice is when you get what you deserve\.
Law is when you get what you pay for\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From victor@vsespb.ru

Ok\, what if I try to rework this poc code as test case\, and submit as patch?

Seems this code fails also on 5\.8\.x \(probably different reason\)\, and
commit be48bbe8d671b6841c3ec7cb734b98071afe3cd9 does not contain test\.

On Wed Jul 31 15&#8203;:40&#8203;:17 2013\, davem wrote&#8203;:
> On Wed\, Jul 31\, 2013 at 12&#8203;:30&#8203;:58PM \-0700\, Victor Efimov wrote&#8203;:
> \(stuff about EINTR\)
> > I wonder if this a bug\, maybe a fix should be backported\, or a test case
> > created\.
> 
> This was fixed in 5\.15\.4 with the following commit\. It's unlikely to to
> be back\-ported to 5\.14\.x\, since that branch is now end\-of\-life\.
> 
> 
> commit be48bbe8d671b6841c3ec7cb734b98071afe3cd9
> Author&#8203;: Chip \<chip@&#8203;pobox\.com>
> Date&#8203;:   Mon Sep 19 23&#8203;:51&#8203;:49 2011 \-0700
> 
>     add a couple missing LEAVEs in perlio\_async\_run\(\)
>

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Jul 31 16&#8203;:10&#8203;:43 2013\, vsespb wrote&#8203;:
> Ok\, what if I try to rework this poc code as test case\, and submit as
patch?
> 
> Seems this code fails also on 5\.8\.x \(probably different reason\)\, and
> commit be48bbe8d671b6841c3ec7cb734b98071afe3cd9 does not contain test\.

If you could\, that would be much appreciated\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From victor@vsespb.ru

patch attached

On Wed Jul 31 17&#8203;:58&#8203;:23 2013\, sprout wrote&#8203;:
> On Wed Jul 31 16&#8203;:10&#8203;:43 2013\, vsespb wrote&#8203;:
> > Ok\, what if I try to rework this poc code as test case\, and submit as
> patch?
> > 
> > Seems this code fails also on 5\.8\.x \(probably different reason\)\, and
> > commit be48bbe8d671b6841c3ec7cb734b98071afe3cd9 does not contain test\.
> 
> If you could\, that would be much appreciated\.
>

**Comment by p5pRT** [comment]
### From victor@vsespb.ru

<details><summary>0001-Print-and-EINTR-test.patch</summary>

```diff
From 39e247562f389ff1f0c27913f782df0d512ccfd9 Mon Sep 17 00:00:00 2001
From: Victor <victor@vsespb.ru>
Date: Fri, 2 Aug 2013 14:39:59 +0400
Subject: [PATCH] Print and EINTR test

Test that print() is not returning EINTR,
fails under 5.14.x ( see https://rt.perl.org/rt3/Ticket/Display.html?id=119097 )
also fails under 5.8.x

Currently test enabled on linux/bsd/solaris/darwin
---
 t/io/eintr_print.t |   87 ++++++++++++++++++++++++++++++++++++++++++++++++++++
 1 files changed, 87 insertions(+), 0 deletions(-)
 create mode 100644 t/io/eintr_print.t

diff --git a/t/io/eintr_print.t b/t/io/eintr_print.t
new file mode 100644
index 0000000..56ab5b4
--- /dev/null
+++ b/t/io/eintr_print.t
@@ -0,0 +1,87 @@
+#!./perl
+
+# print should not return EINTR
+# fails under 5.14.x see https://rt.perl.org/rt3/Ticket/Display.html?id=119097
+# also fails under 5.8.x
+
+BEGIN {
+    chdir 't' if -d 't';
+    @INC = '../lib';
+}
+
+use strict;
+use warnings;
+
+use Config;
+use Time::HiRes;
+use IO::Handle;
+
+require './test.pl';
+
+skip_all("only for dev versions for now") if ((int($]*1000) & 1) == 0);
+skip_all("does not match platform whitelist")
+    unless ($^O =~ /^(linux|.*bsd|darwin|solaris)$/);
+skip_all("ualarm() not implemented on this platform")
+    unless Time::HiRes::d_ualarm();
+skip_all("usleep() not implemented on this platform")
+    unless Time::HiRes::d_usleep();
+skip_all("pipe not implemented on this platform")
+    unless eval { pipe my $in, my $out; 1; };
+
+my $sample = 'abxhrtf6';
+my $full_sample = 'abxhrtf6' x (8192-7);
+my $sample_l = length $full_sample;
+
+my $ppid = $$;
+
+pipe my $in, my $out;
+
+my $small_delay = 10_000;
+my $big_delay = $small_delay * 3;
+my $fail_delay = 20_000_000;
+
+if (my $pid = fork()) {
+    plan(tests => 20);
+
+    local $SIG{ALRM} = sub { print STDERR "FAILED $$\n"; exit(1) };
+    my $child_exited = 0;
+    $in->autoflush(1);
+    $in->blocking(1);
+    binmode $in, ":perlio";
+
+    Time::HiRes::usleep $big_delay;
+
+    # in case test fail it should not hang, however this is not always helping
+    Time::HiRes::ualarm($fail_delay);
+    for (1..10) {
+	my $n = read($in, my $x, $sample_l);
+	die "EOF" unless $n;
+
+	# should return right amount of data
+	is($n, $sample_l);
+
+	# should return right data
+	# don't use "is()" as output in case of fail is big and useless
+	ok($x eq $full_sample);
+    }
+    Time::HiRes::ualarm(0);
+
+    while(wait() != -1 ){};
+} else {
+    local $SIG{ALRM} = sub { print "# ALRM $$\n" };
+    $out->autoflush(1);
+    $out->blocking(1);
+    binmode $out, ":perlio";
+
+    for (1..10) { # on some iteration print() will block
+	Time::HiRes::ualarm($small_delay); # and when it block we'll get SIGALRM
+	# it should unblock and continue after $big_delay
+	die "print failed [ $! ]" unless print($out $full_sample);
+	Time::HiRes::ualarm(0);
+    }
+
+    exit(0);
+}
+
+1;
+
-- 
1.7.0.4


```
</details>

**Comment by p5pRT** [comment]
### From @nwc10

On Fri\, Aug 02\, 2013 at 03&#8203;:43&#8203;:07AM \-0700\, Victor Efimov via RT wrote&#8203;:
> patch attached
> 
> On Wed Jul 31 17&#8203;:58&#8203;:23 2013\, sprout wrote&#8203;:
> > On Wed Jul 31 16&#8203;:10&#8203;:43 2013\, vsespb wrote&#8203;:
> > > Ok\, what if I try to rework this poc code as test case\, and submit as
> > patch?
> > > 
> > > Seems this code fails also on 5\.8\.x \(probably different reason\)\, and
> > > commit be48bbe8d671b6841c3ec7cb734b98071afe3cd9 does not contain test\.
> > 
> > If you could\, that would be much appreciated\.
> > 

I've added the file to MANIFEST\, edited the commit message slightly&#8203;:

commit 4a7b8c665e85a42957110900dd00dd5accaf9e46
Author&#8203;: Victor \<victor@&#8203;vsespb\.ru>
Date&#8203;:   Fri Aug 2 14&#8203;:39&#8203;:59 2013 \+0400

&nbsp;    Test that print\(\) is not returning EINTR\.
&nbsp;    
&nbsp;    fails under 5\.14\.x \( see RT \#119097 \)
&nbsp;    also fails under 5\.8\.x
&nbsp;    
&nbsp;    Currently test enabled on linux/bsd/solaris/darwin

 MANIFEST           |    1 \+
 t/io/eintr\_print\.t |   87 \+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+
 2 files changed\, 88 insertions\(\+\)\, 0 deletions\(\-\)


and pushed it to smoke\-me/nicholas/rt\-119097


I admit that I \*haven't\* actually looked closely at the code\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @tonycoz

On Thu Aug 08 01&#8203;:58&#8203;:52 2013\, nicholas wrote&#8203;:
> On Fri\, Aug 02\, 2013 at 03&#8203;:43&#8203;:07AM \-0700\, Victor Efimov via RT wrote&#8203;:
> > patch attached
> >
> > On Wed Jul 31 17&#8203;:58&#8203;:23 2013\, sprout wrote&#8203;:
> > > On Wed Jul 31 16&#8203;:10&#8203;:43 2013\, vsespb wrote&#8203;:
> > > > Ok\, what if I try to rework this poc code as test case\, and
> submit as
> > > patch?
> > > >
> > > > Seems this code fails also on 5\.8\.x \(probably different reason\)\,
> and
> > > > commit be48bbe8d671b6841c3ec7cb734b98071afe3cd9 does not contain
> test\.
> > >
> > > If you could\, that would be much appreciated\.
> > >
> 
> I've added the file to MANIFEST\, edited the commit message slightly&#8203;:
> 
> commit 4a7b8c665e85a42957110900dd00dd5accaf9e46
> Author&#8203;: Victor \<victor@&#8203;vsespb\.ru>
> Date&#8203;:   Fri Aug 2 14&#8203;:39&#8203;:59 2013 \+0400
> 
>     Test that print\(\) is not returning EINTR\.
> 
>     fails under 5\.14\.x \( see RT \#119097 \)
>     also fails under 5\.8\.x
> 
>     Currently test enabled on linux/bsd/solaris/darwin
> 
>  MANIFEST           |    1 \+
>  t/io/eintr\_print\.t |   87
> \+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+
>  2 files changed\, 88 insertions\(\+\)\, 0 deletions\(\-\)
> 
> 
> and pushed it to smoke\-me/nicholas/rt\-119097
> 
> 
> I admit that I \*haven't\* actually looked closely at the code\.

This test has blocked one of my darwin smokers\, but doesn't appear to
block every time&#8203;:

bash\-3\.2$ ps w
&nbsp;  PID   TT  STAT      TIME COMMAND
 1466 s000  Ss     0&#8203;:00\.02 \-bash
 1788 s000  S\+     0&#8203;:00\.84 screen \-S smoke
 1794 s001  S      0&#8203;:00\.02 bash
 1806 s001  S\+     0&#8203;:00\.53 perl /Users/perlsmoke/bin/smoke\-me\-smoker\.pl
smoke\.cfg
 1899 s001  S\+     0&#8203;:00\.00 sh \-c cd /Users/perlsmoke/smoke\-me/smoke &&
\./smokecurrent\.sh \-nosmartsmoke \-nomail  \</dev/null
 1900 s001  SN\+    0&#8203;:00\.01 /bin/sh \./smokecurrent\.sh \-nosmartsmoke \-nomail
 1902 s001  SN\+    0&#8203;:01\.43 /usr/bin/perl \./smokeperl\.pl \-c
smokecurrent\_config \-nosmartsmoke \-nomail
17935 s001  SN\+    0&#8203;:00\.14 make test\_harness
21640 s001  SN\+    0&#8203;:00\.00 /bin/sh \-e \./runtests choose
21642 s001  SN\+    0&#8203;:40\.22 \./perl harness
22314 s001  SN\+    0&#8203;:00\.04 \./perl \-I\.\. \-MTestInit io/eintr\_print\.t
22315 s001  ZN\+    0&#8203;:00\.00 \(perl\)
58632 s002  S      0&#8203;:00\.02 bash
bash\-3\.2$ \./smokestatus\.pl \-am
Checking status for configuration 'smokecurrent\_config' \(5\.11\.x\)
&nbsp;  Change number 4a7b8c665e85a42957110900dd00dd5accaf9e46 started on Fri
Aug  9 10&#8203;:54&#8203;:54 2013\.
&nbsp;    1 out of 16 configurations finished in 1 day 21 hours 44 minutes\.
&nbsp;    1 configuration showed failures \(F\)\.
&nbsp;    0 failures in the running configuration\.
&nbsp;    15 configurations to finish\, estimated completion in 28 days 13
hours 53 minutes
&nbsp;    Average smoke duration&#8203;: 1 day 21 hours 44 minutes\.
&nbsp;  Matrix\, using cc version 4\.2\.1 \(Based on Apple Inc\. build 5658\) \(LLVM
build 2335\.15\.00\)&#8203;:
&nbsp;  v5\.19\.2\-316\-g4a7b8c6  Configuration \(common\) none
&nbsp;  \-\-\-\-\-\-\-\-\-\-\- \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
&nbsp;  F O ? \-
&nbsp;  | | | \+\-\-\-\-\- PERLIO = perlio \-DDEBUGGING
&nbsp;  | | \+\-\-\-\-\-\-\- PERLIO = stdio  \-DDEBUGGING
&nbsp;  | \+\-\-\-\-\-\-\-\-\- PERLIO = perlio
&nbsp;  \+\-\-\-\-\-\-\-\-\-\-\- PERLIO = stdio
bash\-3\.2$ uname \-a
Darwin pallas\.tony\.develop\-help\.com 11\.4\.2 Darwin Kernel Version 11\.4\.2&#8203;:
Thu Aug 23 16&#8203;:25&#8203;:48 PDT 2012; root&#8203;:xnu\-1699\.32\.7~1/RELEASE\_X86\_64 x86\_64

The F is from me killing the first blockage\.

Running the test directly and via harness in a loop a few hundred times
didn't block\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

On Sat Aug 10 15&#8203;:51&#8203;:13 2013\, tonyc wrote&#8203;:
> This test has blocked one of my darwin smokers\, but doesn't appear to
> block every time&#8203;:

It also blocked on Linux amd64\, Solaris 11\, NetBSD 5\.1\.2\.

Tony

**Comment by p5pRT** [comment]
### From victor@vsespb.ru

hm\, that's sad :\( i've tested it for stability like this on Linux x86\-64
and OpenBSD&#8203;:

\( seq 10000 |xargs \-P 100 \-n 1 \./perl t/io/eintr\_print\.t  \) && echo ALL\_FINE

\(i\.e\. 100 concurrent runs\) and it was fine\.

On Sat Aug 10 17&#8203;:02&#8203;:36 2013\, tonyc wrote&#8203;:
> On Sat Aug 10 15&#8203;:51&#8203;:13 2013\, tonyc wrote&#8203;:
> > This test has blocked one of my darwin smokers\, but doesn't appear to
> > block every time&#8203;:
> 
> It also blocked on Linux amd64\, Solaris 11\, NetBSD 5\.1\.2\.
> 
> Tony

**Comment by p5pRT** [comment]
### From victor@vsespb.ru

Seems it hangs \(in 100% of cases\) when PERLIO=stdio

can be fixed with

skip\_all\("not supposed to work with stdio"\)
&nbsp;    if \(exists $ENV\{PERLIO\} && $ENV\{PERLIO\} =~ /stdio/ \);

similar code exits in eintr\.t&#8203;:

\# XXX for some reason the stdio layer doesn't seem to interrupt
\# write system call when the alarm triggers\.  This makes the tests
\# hang\.

if \(exists $ENV\{PERLIO\} && $ENV\{PERLIO\} =~ /stdio/  \) \{
&nbsp;        skip\_all\('stdio not supported for this script'\);
&nbsp;        exit 0;
\}



On Sun Aug 11 01&#8203;:19&#8203;:12 2013\, vsespb wrote&#8203;:
> hm\, that's sad :\( i've tested it for stability like this on Linux x86\-64
> and OpenBSD&#8203;:
> 
> \( seq 10000 |xargs \-P 100 \-n 1 \./perl t/io/eintr\_print\.t  \) && echo
ALL\_FINE
> 
> \(i\.e\. 100 concurrent runs\) and it was fine\.
> 
> On Sat Aug 10 17&#8203;:02&#8203;:36 2013\, tonyc wrote&#8203;:
> > On Sat Aug 10 15&#8203;:51&#8203;:13 2013\, tonyc wrote&#8203;:
> > > This test has blocked one of my darwin smokers\, but doesn't appear to
> > > block every time&#8203;:
> > 
> > It also blocked on Linux amd64\, Solaris 11\, NetBSD 5\.1\.2\.
> > 
> > Tony
> 
>

**Comment by p5pRT** [comment]
### From victor@vsespb.ru

Also\,

\(1\)

http&#8203;://perldoc\.perl\.org/perlipc\.html\#Deferred\-Signals\-%28Safe\-Signals%29
> the solution is to use the :perlio layer to do IO\-\-at least on those
handles that you want to be able to break into with signals\.

http&#8203;://perldoc\.perl\.org/PerlIO\.html
> The default can be overridden by setting the environment variable
PERLIO to a space separated list of layers

it seems that the test initially had
 binmode $fh\, "&#8203;:perlio";

but PERLIO=stdio overrides this\.
so phrase "The default can be overridden by setting the environment
variable PERLIO" does not look correct\. Not only default is overriden\,
but any atempt to use the layer\.

\(2\)

There is a ticket related to eintr\.t
https://rt-archive.perl.org/perl5/Ticket/Display.html?id=85842
Long story short \(my understanding\)&#8203;: 

1\. Commit http&#8203;://perl5\.git\.perl\.org/perl\.git/commit/b83080de5c4254
introduce eintr\.t problems under freebsd and some solaris versions
2\. Those OS blacklisted in eintr\.t
3\. It still has problems on older linuxes
4\. According to this comment
https://rt-archive.perl.org/perl5/Ticket/Display.html?id=84688#txn-871204
any \(or some\) IO in signal handler can be a problem\, however it should
not \(according to perlipc\)

So\, according to this ticket it might be a good idea to&#8203;:
a\) copy OS blacklist from eintr\.t to eintr\_print\.t
b\) remove 'print "\# ALRM $$\\n"' from $SIG\{ALRM\}




On Sun Aug 11 02&#8203;:41&#8203;:44 2013\, vsespb wrote&#8203;:
> Seems it hangs \(in 100% of cases\) when PERLIO=stdio
> 
> can be fixed with
> 
> skip\_all\("not supposed to work with stdio"\)
>     if \(exists $ENV\{PERLIO\} && $ENV\{PERLIO\} =~ /stdio/ \);
> 
> similar code exits in eintr\.t&#8203;:
> 
> \# XXX for some reason the stdio layer doesn't seem to interrupt
> \# write system call when the alarm triggers\.  This makes the tests
> \# hang\.
> 
> if \(exists $ENV\{PERLIO\} && $ENV\{PERLIO\} =~ /stdio/  \) \{
>         skip\_all\('stdio not supported for this script'\);
>         exit 0;
> \}
> 
> 
> 
> On Sun Aug 11 01&#8203;:19&#8203;:12 2013\, vsespb wrote&#8203;:
> > hm\, that's sad :\( i've tested it for stability like this on Linux x86\-64
> > and OpenBSD&#8203;:
> > 
> > \( seq 10000 |xargs \-P 100 \-n 1 \./perl t/io/eintr\_print\.t  \) && echo
> ALL\_FINE
> > 
> > \(i\.e\. 100 concurrent runs\) and it was fine\.
> > 
> > On Sat Aug 10 17&#8203;:02&#8203;:36 2013\, tonyc wrote&#8203;:
> > > On Sat Aug 10 15&#8203;:51&#8203;:13 2013\, tonyc wrote&#8203;:
> > > > This test has blocked one of my darwin smokers\, but doesn't
appear to
> > > > block every time&#8203;:
> > > 
> > > It also blocked on Linux amd64\, Solaris 11\, NetBSD 5\.1\.2\.
> > > 
> > > Tony
> > 
> > 
> 
>

**Comment by p5pRT** [comment]
### From @Leont

On Sun\, Aug 11\, 2013 at 3&#8203;:18 PM\, Victor Efimov via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> Also\,
>
> \(1\)
>
> http&#8203;://perldoc\.perl\.org/perlipc\.html\#Deferred\-Signals\-%28Safe\-Signals%29
>> the solution is to use the :perlio layer to do IO\-\-at least on those
> handles that you want to be able to break into with signals\.
>
> http&#8203;://perldoc\.perl\.org/PerlIO\.html
>> The default can be overridden by setting the environment variable
> PERLIO to a space separated list of layers
>
> it seems that the test initially had
>  binmode $fh\, "&#8203;:perlio";

binmode $fh\, "&#8203;:perlio" is conceptually nonsensical\. You want to
specify that during open\, but that's broken too :\-/

> but PERLIO=stdio overrides this\.
> so phrase "The default can be overridden by setting the environment
> variable PERLIO" does not look correct\. Not only default is overriden\,
> but any atempt to use the layer\.

No\, it doesn't\. What you're doing is effectively both&#8203;:
"&#8203;:stdio&#8203;:perlio"\. This will generally block unless you're reading the
buffer size\(Because :perlio is assuming read sementics underneath it\,
not fread semantics\)\. This is a known issue that can't be fixed \(stdio
just sucks like that\)\.

You can kind of fix it with a «use open IO => "&#8203;:pop&#8203;:perlio";»\, it
isn't pretty but I'm not sure what else would do the trick\. PerlIO
sucks like that\.

Leon

**Comment by p5pRT** [comment]
### From victor@vsespb.ru

ok\, so easier just to

skip\_all\("not supposed to work with stdio"\)
if \(exists $ENV\{PERLIO\} && $ENV\{PERLIO\} =~ /stdio/ \);

like eintr\.t does

On Sun Aug 11 05&#8203;:55&#8203;:55 2013\, LeonT wrote&#8203;:
> On Sun\, Aug 11\, 2013 at 3&#8203;:18 PM\, Victor Efimov via RT
> \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > Also\,
> >
> > \(1\)
> >
> > http&#8203;://perldoc\.perl\.org/perlipc\.html\#Deferred\-Signals\-%28Safe\-Signals%29
> >> the solution is to use the :perlio layer to do IO\-\-at least on those
> > handles that you want to be able to break into with signals\.
> >
> > http&#8203;://perldoc\.perl\.org/PerlIO\.html
> >> The default can be overridden by setting the environment variable
> > PERLIO to a space separated list of layers
> >
> > it seems that the test initially had
> >  binmode $fh\, "&#8203;:perlio";
> 
> binmode $fh\, "&#8203;:perlio" is conceptually nonsensical\. You want to
> specify that during open\, but that's broken too :\-/
> 
> > but PERLIO=stdio overrides this\.
> > so phrase "The default can be overridden by setting the environment
> > variable PERLIO" does not look correct\. Not only default is overriden\,
> > but any atempt to use the layer\.
> 
> No\, it doesn't\. What you're doing is effectively both&#8203;:
> "&#8203;:stdio&#8203;:perlio"\. This will generally block unless you're reading the
> buffer size\(Because :perlio is assuming read sementics underneath it\,
> not fread semantics\)\. This is a known issue that can't be fixed \(stdio
> just sucks like that\)\.
> 
> You can kind of fix it with a �use open IO => "&#8203;:pop&#8203;:perlio";�\, it
> isn't pretty but I'm not sure what else would do the trick\. PerlIO
> sucks like that\.
> 
> Leon
>

**Comment by p5pRT** [comment]
### From victor@vsespb.ru

attaching fixes to eintr\_print\.t

On Sun Aug 11 05&#8203;:18&#8203;:43 2013\, vsespb wrote&#8203;:
> Also\,
> 
> \(1\)
> 
> http&#8203;://perldoc\.perl\.org/perlipc\.html\#Deferred\-Signals\-%28Safe\-Signals%29
> > the solution is to use the :perlio layer to do IO\-\-at least on those
> handles that you want to be able to break into with signals\.
> 
> http&#8203;://perldoc\.perl\.org/PerlIO\.html
> > The default can be overridden by setting the environment variable
> PERLIO to a space separated list of layers
> 
> it seems that the test initially had
>  binmode $fh\, "&#8203;:perlio";
> 
> but PERLIO=stdio overrides this\.
> so phrase "The default can be overridden by setting the environment
> variable PERLIO" does not look correct\. Not only default is overriden\,
> but any atempt to use the layer\.
> 
> \(2\)
> 
> There is a ticket related to eintr\.t
> https://rt-archive.perl.org/perl5/Ticket/Display.html?id=85842
> Long story short \(my understanding\)&#8203;: 
> 
> 1\. Commit http&#8203;://perl5\.git\.perl\.org/perl\.git/commit/b83080de5c4254
> introduce eintr\.t problems under freebsd and some solaris versions
> 2\. Those OS blacklisted in eintr\.t
> 3\. It still has problems on older linuxes
> 4\. According to this comment
> https://rt-archive.perl.org/perl5/Ticket/Display.html?id=84688#txn-871204
> any \(or some\) IO in signal handler can be a problem\, however it should
> not \(according to perlipc\)
> 
> So\, according to this ticket it might be a good idea to&#8203;:
> a\) copy OS blacklist from eintr\.t to eintr\_print\.t
> b\) remove 'print "\# ALRM $$\\n"' from $SIG\{ALRM\}
> 
> 
> 
> 
> On Sun Aug 11 02&#8203;:41&#8203;:44 2013\, vsespb wrote&#8203;:
> > Seems it hangs \(in 100% of cases\) when PERLIO=stdio
> > 
> > can be fixed with
> > 
> > skip\_all\("not supposed to work with stdio"\)
> >     if \(exists $ENV\{PERLIO\} && $ENV\{PERLIO\} =~ /stdio/ \);
> > 
> > similar code exits in eintr\.t&#8203;:
> > 
> > \# XXX for some reason the stdio layer doesn't seem to interrupt
> > \# write system call when the alarm triggers\.  This makes the tests
> > \# hang\.
> > 
> > if \(exists $ENV\{PERLIO\} && $ENV\{PERLIO\} =~ /stdio/  \) \{
> >         skip\_all\('stdio not supported for this script'\);
> >         exit 0;
> > \}
> > 
> > 
> > 
> > On Sun Aug 11 01&#8203;:19&#8203;:12 2013\, vsespb wrote&#8203;:
> > > hm\, that's sad :\( i've tested it for stability like this on Linux
x86\-64
> > > and OpenBSD&#8203;:
> > > 
> > > \( seq 10000 |xargs \-P 100 \-n 1 \./perl t/io/eintr\_print\.t  \) && echo
> > ALL\_FINE
> > > 
> > > \(i\.e\. 100 concurrent runs\) and it was fine\.
> > > 
> > > On Sat Aug 10 17&#8203;:02&#8203;:36 2013\, tonyc wrote&#8203;:
> > > > On Sat Aug 10 15&#8203;:51&#8203;:13 2013\, tonyc wrote&#8203;:
> > > > > This test has blocked one of my darwin smokers\, but doesn't
> appear to
> > > > > block every time&#8203;:
> > > > 
> > > > It also blocked on Linux amd64\, Solaris 11\, NetBSD 5\.1\.2\.
> > > > 
> > > > Tony
> > > 
> > > 
> > 
> > 
> 
>

**Comment by p5pRT** [comment]
### From victor@vsespb.ru

<details><summary>0025-Fixing-eintr_print.t-intermittent-hang.patch</summary>

```diff
From ea8aff5fe307d300ce2a72120f941102fe0dd016 Mon Sep 17 00:00:00 2001
From: Victor <victor@vsespb.ru>
Date: Mon, 12 Aug 2013 12:49:58 +0400
Subject: [PATCH 25/25] Fixing eintr_print.t intermittent hang

1. Disable test for PERLIO=stdio
2. Remove binmode - it turns out it's useless
3. Copy OS blacklist from eintr.t ( RT #85842, RT #84688)
4. Add additional delay before child exit, just in case.
---
 t/io/eintr_print.t |   14 +++++++++++---
 1 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/t/io/eintr_print.t b/t/io/eintr_print.t
index 56ab5b4..1e378fa 100644
--- a/t/io/eintr_print.t
+++ b/t/io/eintr_print.t
@@ -27,6 +27,16 @@ skip_all("usleep() not implemented on this platform")
     unless Time::HiRes::d_usleep();
 skip_all("pipe not implemented on this platform")
     unless eval { pipe my $in, my $out; 1; };
+skip_all("not supposed to work with stdio")
+    if (defined $ENV{PERLIO} && $ENV{PERLIO} =~ /stdio/ );
+
+# copy OS blacklist from eintr.t ( related to perl #85842 and #84688 )
+my ($osmajmin) = $Config{osvers} =~ /^(\d+\.\d+)/;
+
+skip_all('various portability issues')
+    if ( $^O =~ /freebsd/ || $^O eq 'midnightbsd' ||
+	($^O eq 'solaris' && $Config{osvers} eq '2.8') ||
+	($^O eq 'darwin' && $osmajmin < 9) );
 
 my $sample = 'abxhrtf6';
 my $full_sample = 'abxhrtf6' x (8192-7);
@@ -47,7 +57,6 @@ if (my $pid = fork()) {
     my $child_exited = 0;
     $in->autoflush(1);
     $in->blocking(1);
-    binmode $in, ":perlio";
 
     Time::HiRes::usleep $big_delay;
 
@@ -71,7 +80,6 @@ if (my $pid = fork()) {
     local $SIG{ALRM} = sub { print "# ALRM $$\n" };
     $out->autoflush(1);
     $out->blocking(1);
-    binmode $out, ":perlio";
 
     for (1..10) { # on some iteration print() will block
 	Time::HiRes::ualarm($small_delay); # and when it block we'll get SIGALRM
@@ -79,7 +87,7 @@ if (my $pid = fork()) {
 	die "print failed [ $! ]" unless print($out $full_sample);
 	Time::HiRes::ualarm(0);
     }
-
+    Time::HiRes::usleep(500_000);
     exit(0);
 }
 
-- 
1.7.0.4


```
</details>

**Comment by p5pRT** [comment]
### From @tonycoz

On Mon Aug 12 02&#8203;:01&#8203;:50 2013\, vsespb wrote&#8203;:
> attaching fixes to eintr\_print\.t

Applied as b893ae5e183df130e3225a0aaf8ac963a6f6d1ff \(the original
patch\)\, 878bfcdfc7542cbd6624b88ea6e16bd3ac01b3cd \(the skips\) and a merge
commit 18603efee1e80bda2f6107c9569d056cb1071d4b to bring them all and in
the darkness\.\.\. urr\, to combine them\.

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'resolved'
